File: examples/python/ttl-caching/simple_crud.py
#!/usr/bin/env python3
"""Simple CRUD Examples - Following benchmark patterns"""

import asyncio
import os
import json
from pathlib import Path
import aiohttp
from dotenv import load_dotenv

# Load environment variables
env_path = Path(__file__).parent.parent.parent / ".env"
load_dotenv(env_path)

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")


async def get_auth_token(session):
    """Get authentication token"""
    async with session.post(
        f"{BASE_URL}/api/auth/token",
        json={"api_key": API_KEY},
        headers={"Content-Type": "application/json"},
    ) as response:
        data = await response.json()
        return data["token"]


async def simple_crud_examples():
    """Run simple CRUD examples"""
    async with aiohttp.ClientSession() as session:
        # Get authentication token
        token = await get_auth_token(session)
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        }
        print("✓ Authentication successful")

        try:
            # Example 1: Insert a document
            print("\n=== Insert Document ===")
            async with session.post(
                f"{BASE_URL}/api/insert/test_collection",
                headers=headers,
                json={"name": "Test Record", "value": 42, "active": True},
            ) as response:
                doc = await response.json()
                print(f"Inserted: {doc}")
                doc_id = doc["id"]

            # Example 2: Find by ID
            print("\n=== Find by ID ===")
            async with session.get(
                f"{BASE_URL}/api/find/test_collection/{doc_id}",
                headers={"Authorization": f"Bearer {token}"},
            ) as response:
                found_doc = await response.json()
                print(f"Found: {found_doc}")

            # Example 3: Find with query
            print("\n=== Find with Query ===")
            async with session.post(
                f"{BASE_URL}/api/find/test_collection",
                headers=headers,
                json={
                    "filter": {
                        "type": "Condition",
                        "content": {"field": "active", "operator": "Eq", "value": True},
                    },
                    "limit": 10,
                },
            ) as response:
                docs = await response.json()
                print(
                    f'Found documents: {len(docs) if isinstance(docs, list) else "N/A"}'
                )

            # Example 4: Update document
            print("\n=== Update Document ===")
            async with session.put(
                f"{BASE_URL}/api/update/test_collection/{doc_id}",
                headers=headers,
                json={"name": "Updated Record", "value": 100},
            ) as response:
                updated = await response.json()
                print(f"Updated: {updated}")

            # Example 5: Delete document
            print("\n=== Delete Document ===")
            async with session.delete(
                f"{BASE_URL}/api/delete/test_collection/{doc_id}",
                headers={"Authorization": f"Bearer {token}"},
            ) as response:
                print("Deleted document")

            print("\n✓ All CRUD operations completed successfully")

        except Exception as e:
            print(f"Error: {e}")
            raise


if __name__ == "__main__":
    asyncio.run(simple_crud_examples())

File: examples/javascript/ttl-caching/simple_crud.js
// Simple CRUD Examples - Following benchmark patterns
const fetch = require('node-fetch');
require('dotenv').config({ path: require('path').join(__dirname, '..', '..', '.env') });

async function simpleCRUDExamples() {
    const BASE_URL = process.env.API_BASE_URL || 'http://localhost:8080';
    const API_KEY = process.env.API_BASE_KEY || 'a-test-api-key-from-ekodb';
    
    // Get authentication token
    let TOKEN = process.env.AUTH_TOKEN;
    if (!TOKEN) {
        const tokenResponse = await fetch(`${BASE_URL}/api/auth/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ api_key: API_KEY })
        });
        const tokenData = await tokenResponse.json();
        TOKEN = tokenData.token;
    }

    console.log('✓ Authentication successful');

    try {
        // Example 1: Insert a document
        console.log('\n=== Insert Document ===');
        const insertResponse = await fetch(`${BASE_URL}/api/insert/test_collection`, {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${TOKEN}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                name: 'Test Record',
                value: 42,
                active: true
            })
        });
        const doc = await insertResponse.json();
        console.log('Inserted:', doc);

        // Example 2: Find by ID
        console.log('\n=== Find by ID ===');
        const findByIdResponse = await fetch(`${BASE_URL}/api/find/test_collection/${doc.id}`, {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${TOKEN}`
            }
        });
        const foundDoc = await findByIdResponse.json();
        console.log('Found:', foundDoc);

        // Example 3: Find with query
        console.log('\n=== Find with Query ===');
        const findResponse = await fetch(`${BASE_URL}/api/find/test_collection`, {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${TOKEN}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                filter: {
                    type: 'Condition',
                    content: {
                        field: 'active',
                        operator: 'Eq',
                        value: true
                    }
                },
                limit: 10
            })
        });
        const docs = await findResponse.json();
        console.log('Found documents:', docs.length);

        // Example 4: Update document
        console.log('\n=== Update Document ===');
        const updateResponse = await fetch(`${BASE_URL}/api/update/test_collection/${doc.id}`, {
            method: 'PUT',
            headers: {
                'Authorization': `Bearer ${TOKEN}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                name: 'Updated Record',
                value: 100
            })
        });
        const updated = await updateResponse.json();
        console.log('Updated:', updated);

        // Example 5: Delete document
        console.log('\n=== Delete Document ===');
        await fetch(`${BASE_URL}/api/delete/test_collection/${doc.id}`, {
            method: 'DELETE',
            headers: {
                'Authorization': `Bearer ${TOKEN}`
            }
        });
        console.log('Deleted document');

        console.log('\n✓ All CRUD operations completed successfully');

    } catch (error) {
        console.error('Error:', error);
        process.exit(1);
    }
}

// Run the examples
simpleCRUDExamples().catch((error) => {
  console.error("❌ Error:", error);
  process.exit(1);
});
